### PR TITLE
Juke Build

### PR DESCRIPTION
**Pull request by stylemistake**
## About The Pull Request

image

Integrates [Juke Build](https://github.com/stylemistake/juke-build) into CBT.

This was an effort to extract CBT functionality into an external repository/package with a refined API and extra features.

## Important

**tgui** and **tgfont** assets no longer block the compilation of DME. They are loaded dynamically on server startup. The side effect of this, is that tgui bundle errors will manifest in Dream Daemon instead of the compiler.

Compiler still shows a warning if you compile DME directly, so hopefully there will be no surprise.

## Features of Juke Build

- Simple and strongly typed API
- Robust dependency model with concurrent execution of targets
- CLI argument (and environment) parsing via Parameters API
- Nice colorized output (at least on Linux)

## Features of CBT (new)

- You can specify which specific target to build via command line arguments.
- You can specify defines by stacking `-Ddefinename` flags, or by using `DEFINE=A,B,C` env var with comma separated values.
  - When using flags, order of flags and task names matters (more in [Juke Build docs](https://github.com/stylemistake/juke-build)). Tell me if this is something you all are ok with.
- Parallel execution drastically cuts down build time. It's pretty much the same as if you compiled just dm and nothing else.

## Removals

Removed CBT_BUILD_MODE env var, in favor of an optional `tgs` target, which you can launch via `tools/build/build tgs`, and in all other cases by using the `-D` flags to add defines.

## Did I test it?

Juke Build itself has a tiny unit test, and each piece of the CI pipeline was tested in isolation on my machine.

Not sure what bugs will pop up in practice.

CI integration tests might not need to build TGUI now, but I left that job as is.

## Screenshots

image

### CODE REVIEW

**Comment by YakumoChen** [comment]
Holy hell, this sounds based. I might actually be able to code on my absolute trash heap PC without asking someone for a TGUI compile.

**Review by Mothblocks — APPROVED** [review]
I like it, but I'd like to wait to hear from @Cyberboss to see if this has any TGS complications (though I don't think it would).

**Comment by Cyberboss** [comment]
Only complication is that whenever we change a tools/tgs4_scripts file we need to coordinate with ops to update the server. Those files are not updated automatically in the TGS instance.

File: code/modules/asset_cache/asset_list_items.dm [hunk 1]
```diff
@@ -3,21 +3,21 @@
...
+		"tgui-common.bundle.js" = file("tgui/public/tgui-common.bundle.js"),
 	)
 
 /datum/asset/simple/tgui
 	keep_local_name = TRUE
 	assets = list(
-		"tgui.bundle.js" = 'tgui/public/tgui.bundle.js',
-		"tgui.bundle.css" = 'tgui/public/tgui.bundle.css',
+		"tgui.bundle.js" = file("tgui/public/tgui.bundle.js"),
+		"tgui.bundle.css" = file("tgui/public/tgui.bundle.css"),
```
Comment (Cyberboss) [inline code/modules/asset_cache/asset_list_items.dm:13]: tools/deploy.sh needs to be update to include these

Comment (stylemistake) [reply code/modules/asset_cache/asset_list_items.dm:13]: Done.

**Comment by Mothblocks** [comment]
@Cyberboss Has it all been settled?

**Comment by JJRcop** [comment]
AHEM #58245 was nuked with this change

needs hotfix 3

**Pull request merged**